src/editortoolkit_neume.cpp

```diff
@@ -729,80 +894,98 @@ bool EditorToolkitNeume::Insert(std::string elementType, std::string staffId, in
         Surface *surface = dynamic_cast<Surface *>(facsimile->GetFirst(SURFACE));
         surface->AddChild(zone);
         custos->SetZone(zone);
-        custos->SetFacs(zone->GetUuid());
         layer->AddChild(custos);
-        // Find closest valid clef
-        Clef *clef = NULL;
-        clef = layer->GetClef(custos);
-        if (clef == NULL) {
-            LogError("There is no valid clef available.");
-            delete custos;
-            return false;
-        }
-
-        custos->SetOct(3);
-        if (clef->GetShape() == CLEFSHAPE_C)
-            custos->SetPname(PITCHNAME_c);
-        else if (clef->GetShape() == CLEFSHAPE_F)
-            custos->SetPname(PITCHNAME_f);
 
-        const int staffSize = m_doc->GetDrawingUnit(staff->m_drawingStaffSize);
         const int noteHeight = (int)(m_doc->GetDrawingDoubleUnit(staff->m_drawingStaffSize) / 2);
         const int noteWidth = (int)(m_doc->GetDrawingDoubleUnit(staff->m_drawingStaffSize) / 1.4);
-        const int pitchDifference = round(
-            (double)(staff->GetZone()->GetUly() + (2 * staffSize * (staff->m_drawingLines - clef->GetLine())) - (uly))
-            / (double)(staffSize));
 
-        custos->AdjustPitchByOffset(pitchDifference);
         ulx -= noteWidth / 2;
         uly -= noteHeight / 2;
 
         zone->SetUlx(ulx);
         zone->SetUly(uly);
         zone->SetLrx(ulx + noteWidth);
         zone->SetLry(uly + noteHeight);
-        m_editInfo = custos->GetUuid();
+        layer->ReorderByXPos();
+        if (!AdjustPitchFromPosition(custos)) {
+            LogError("Failed to set pitch.");
+
+            m_infoObject.import("status", "FAILURE");
+            m_infoObject.import("message", "Failed to set pitch.");
+            return false;
+        }
+        m_infoObject.import("uuid", custos->GetUuid());
     }
     else {
         LogError("Unsupported type '%s' for insertion", elementType.c_str());
+        m_infoObject.import("status", "FAILURE");
+        m_infoObject.import("message", "Unsupported type '" + elementType + "' for insertion.");
         return false;
     }
     layer->ReorderByXPos();
+    m_infoObject.import("status", status);
+    m_infoObject.import("message", message);
     return true;
 }
 
 bool EditorToolkitNeume::Merge(std::vector<std::string> elementIds)
 {
-    m_editInfo = "";
+    m_infoObject.reset();
```

Comment:
If this is getting called by every edit action individually might it be worthwhile to have it be called by the wrapper functions or in some more built-in way?

Reply:
Yeah this could probably be moved to the `ParseEditorAction` function and then failures in parsing could be included in the info object as well.

Reply:
Addressed this in 1c0cb06